src/query/api/v1/handler/influxdb/write.go

```diff
@@ -0,0 +1,255 @@
+		case imodels.Boolean:
+			v, err := it.BooleanValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding boolean: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			if v {
+				value = 1.0
+			}
+		case imodels.Integer:
+			v, err := it.IntegerValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding integer: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Unsigned:
+			v, err := it.UnsignedValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding unsigned: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Float:
+			v, err := it.FloatValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding float: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = v
+		default:
+			// TBD if we should stick strings as
+			// tags or not; to prevent cardinality
+			// explosion, we drop them for now
+			continue
+		}
+		nlen := len(point.Name())
+		name := make([]byte, nlen+1+len(it.FieldKey()))
+		copy(name, point.Name())
+		copy(name[nlen:], []byte("_"))
+		copy(name[nlen+1:], it.FieldKey())
+
+		name = self.promRewriter.Metric.ToValid(name)
+		self.fields = append(self.fields, &ingestField{name: name, value: value})
+	}
+	return true
+}
+
+func (self *ingestIterator) Next() bool {
+	for len(self.points) > self.pointIndex {
+		if self.nextFieldIndex == 0 {
+			// Populate tags only if we have fields we care about
+			if self.populateFields() {
+				point := self.points[self.pointIndex]
+				ptags := point.Tags()
+				tags := models.NewTags(len(ptags), nil)
+				for _, tag := range ptags {
+					name := self.promRewriter.Label.ToValid(tag.Key)
+					tags = tags.AddTagWithoutNormalizing(models.Tag{Name: name, Value: tag.Value})
+					// Normalization happens when we stick
+					// in __name__ in Current()
+				}
+				self.tags = tags
+			}
+		}
+		self.nextFieldIndex += 1
+		if self.nextFieldIndex > len(self.fields) {
+			self.pointIndex += 1
+			self.nextFieldIndex = 0
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func (self *ingestIterator) Current() (models.Tags, ts.Datapoints, xtime.Unit, []byte) {
+	if self.pointIndex < len(self.points) && self.nextFieldIndex > 0 && len(self.fields) > (self.nextFieldIndex-1) {
+		point := self.points[self.pointIndex]
+		field := self.fields[self.nextFieldIndex-1]
+		tags := self.tags.AddTag(models.Tag{Name: []byte("__name__"),
+			Value: field.name})
+
+		return tags, []ts.Datapoint{ts.Datapoint{Timestamp: point.Time(),
+			Value: field.value}}, xtime.Nanosecond, nil
+	}
+	return models.EmptyTags(), nil, 0, nil
+}
+
+func (self *ingestIterator) Reset() error {
+	self.pointIndex = 0
+	self.nextFieldIndex = 0
+	self.err = xerrors.NewMultiError()
+	return nil
+}
+
+func (self *ingestIterator) Error() error {
+	return self.err.FinalError()
+}
+
+func NewInfluxWriterHandler(downsamplerAndWriter ingest.DownsamplerAndWriter,
```

Comment:
Sorry, was holding off on reviewing this since was hoping to get #2073 in first which refactors how we set up handlers a little (now this should just take a `opts options.HandlerOptions` and then take the downsampler and iOpts from there)

Reply:
Adjusted to using it; now the API looks much cleaner at least.